virtualBridge.py

```diff
@@ -70,6 +74,56 @@ def setup_dhcp_entry(self, cfg: NodeConfig) -> None:
         cmd = f"virsh net-update default add ip-dhcp-host \"{host_xml}\" --live --config"
         self.hostconn.run_or_die(cmd)
 
+    def remove_dhcp_entries(self, vms: list[NodeConfig]) -> None:
+        def filter_dhcp_leases(j: list[dict[str, str]], removed_macs: list[str], names: list[str]) -> list[dict[str, str]]:
+            filtered = []
+            for entry in j:
+                if entry["mac-address"] in removed_macs:
+                    logger.info(f'Removed host with mac {entry["mac-address"]}')
+                    continue
+                if "hostname" in entry and entry["hostname"] in names:
+                    logger.info(f'Removed host with name {entry["hostname"]}')
+                    continue
+                logger.info(f'Kept entry {entry}')
+                filtered.append(entry)
+            return filtered
+
+        xml_str = self.hostconn.run("virsh net-dumpxml default").out
+        q = et.fromstring(xml_str)
+        removed_macs = []  # type: list[str]
+        names = [vm.name for vm in vms]
+        ips = [vm.ip for vm in vms]
+        ip_tree = next((it for it in q.iter("ip")), et.Element(''))
+        dhcp = next((it for it in ip_tree.iter("dhcp")), et.Element(''))
+        for e in dhcp:
+            if e.get('name') in names or e.get('ip') in ips:
+                # For all dhcp entries, check whether the name or the ip has been assigned to one of our "vms", and remove it if it's the case.
+                mac = e.attrib["mac"]
+                name = e.attrib["name"]
+                ip = e.attrib["ip"]
+                pre = "virsh net-update default delete ip-dhcp-host"
+                cmd = f"{pre} \"<host mac='{mac}' name='{name}' ip='{ip}'/>\" --live --config"
+                logger.info(self.hostconn.run(cmd))
+                removed_macs.append(mac)
+
+        fn = "/var/lib/libvirt/dnsmasq/virbr0.status"
+        p = Path(fn)
+        with p.open() as f:
+            contents = f.read()
+
+        if contents:
+            j = json.loads(contents)
+            names = [vm.name for vm in vms]
+            logger.info(f'Cleaning up {fn}')
+            logger.info(f'removing hosts with mac in {removed_macs} or name in {names}')
+            filtered = filter_dhcp_leases(j, removed_macs, names)
```

Comment:
here we compute `names` redundantly from `j`, then pass both on to `filter_dhcp_leases()`. That is ugly. It begs the immediate question, whether the caller made sure that the information is consistent.

On the other hand, at this place, the caller does almost nothing with `j` or `names` except logging. Note that `filter_dhcp_leases()` already does logging too, so there is no reason to have these preparation steps that the caller performs.

In short, move mode code to `filter_dhcp_leases()`. In particular also the logging (it already does logging) and the handling for `names`.

Reply:
a unit test for `filter_dhcp_leases()` would be nice.

Reply:
I do not think that names and j are redundant.
filter_dhcp_leases keep all leases:
- not in removed_macs (which contains the list of removed dhcp entries)
- not in names (which contains the list of names of nodes for which we want to remove the lease).

My understanding (might be wrong) is that the dhcp server might serve other clusters as well, so names (i.e. the nodes for which we want to remove leases) might not cover all leases in j.
Also, filter_dhcp_leases() can (and will in further commits) be used were "names" (vms) only refers to some (and not all) nodes e.g. we want to teardown only one node of a big cluster.

I'll send updated PR for moving more code to filter_dhcp_leases.
Thanks again for the review.

Reply:
> I do not think that names and j are redundant.

You are right! Sorry